scripts/gb_model/compose_network.py

```diff
@@ -486,6 +488,140 @@ def add_EV_DSR_V2G(
     )
 
 
+def _add_dsr_pypsa_components(
+    n: pypsa.Network, df: pd.DataFrame, dsr_hours: list[int], key: str
+):
+    """
+    Add DSR components for a given sector to PyPSA network
+
+    Parameters
+    ----------
+        n : pypsa.Network
+            Network to finalize
+        df : pd.DataFrame
+            DataFrame containing flexibility p_nom data indexed by bus
+        dsr_hours : list[int]
+            Hours during which demand-side management can occur
+        key : str
+            Sector key (e.g., 'residential', 'iandc', 'iandc_heat')
+    """
+
+    # Add the DSR carrier to the PyPSA network
+    n.add(
+        "Carrier",
+        f"{key} DSR",
+        nice_name=f"{key} Demand Side Response",
+    )
+
+    # Add the DSR shift and reverse carriers to the PyPSA network
+    n.add(
+        "Carrier",
+        f"{key} DSR shift",
+    )
+
+    n.add(
+        "Carrier",
+        f"{key} DSR reverse",
+    )
+
+    # Create DSR buses, links and stores
+    # Add the DSR bus to the PyPSA network
+    n.add(
+        "Bus",
+        df.index,
+        suffix=f" {key} DSR bus",
+        carrier=f"{key} DSR",
+        x=n.buses.loc[df.index].x,
+        y=n.buses.loc[df.index].y,
+        country=n.buses.loc[df.index].country,
+    )
+
+    # Add the DSR link from AC bus to DSR bus to the PyPSA network
+    n.add(
+        "Link",
+        df.index,
+        suffix=f" {key} DSR",
+        bus0=df.index,
+        bus1=df.index + f" {key} DSR bus",
+        p_nom=df.p_nom.abs(),
+        efficiency=1.0,
+        carrier=f"{key} DSR shift",
+    )
+
+    # Add the DSR link from DSR bus to AC bus to the PyPSA network
+    n.add(
+        "Link",
+        df.index,
+        suffix=f" {key} DSR reverse",
+        bus0=df.index + f" {key} DSR bus",
+        bus1=df.index,
+        p_nom=df.p_nom.abs(),
+        efficiency=1.0,
+        carrier=f"{key} DSR reverse",
+    )
+
+    breakpoint()
+    # Create DSM profile to set as e_max_pu for DSR store
+    dsr_profile = pd.DataFrame(index=n.snapshots, columns=df.index, data=0.0)
+    mask = (dsr_profile.index.hour >= dsr_hours[0]) & (
+        dsr_profile.index.hour < dsr_hours[1]
+    )
+    dsr_profile.loc[mask] = 1.0
+    EU_columns = [col for col in df.index if "GB" not in col]
+    # Shift European neighbour columns by 1 hour to account for time zone difference
```

Comment:
Not quite so simple since we may include other countries beyond those that are +1hr out. You can use `pytz` to get the timezone and then convert the timeseries index for each non-GB country